Description: Avoid leaking user names.
Bug: http://bugs.debian.org/535877
Author: Dmitry Butskoy

When we detect that it is the first password prompt in the PAM chain,
we always use the standard "Password: " prompt, else we use the
pam_ssh's prompt if it is needed.

CVE-2009-1273

Index: libpam-ssh/pam_get_pass.c
===================================================================
--- libpam-ssh.orig/pam_get_pass.c	2009-07-08 22:46:47.000000000 +0200
+++ libpam-ssh/pam_get_pass.c	2009-07-08 22:49:34.000000000 +0200
@@ -78,13 +78,20 @@
 	const void *item = NULL;
 
 	/*
-	 * Grab the already-entered password if we might want to use it.
+	 * Grab the already-entered password.
 	 */
+	retval = pam_get_item(pamh, PAM_AUTHTOK, &item);
+	/*
+	 * Always use standard prompt for the first time.
+	 */
+	if (item == NULL)
+		prompt = "Password: ";
 	if (pam_test_option(options, PAM_OPT_TRY_FIRST_PASS, NULL) ||
 	    pam_test_option(options, PAM_OPT_USE_FIRST_PASS, NULL)) {
-		retval = pam_get_item(pamh, PAM_AUTHTOK, &item);
 		if (retval != PAM_SUCCESS)
 			return retval;
+	} else {
+		item = NULL;	/* Need own password. */
 	}
 
 	if (item == NULL) {
